helper.go

```diff
@@ -305,7 +305,20 @@ func marshalInt32(value int) (rs []byte, err error) {
 // Counter32, Gauge32, TimeTicks, Unsigned32
 func marshalUint32(v interface{}) ([]byte, error) {
 	bs := make([]byte, 4)
-	source := v.(uint32)
+
+	var source uint32
+	switch val := v.(type) {
+	case uint32:
+		source = val
+	case uint:
+		source = uint32(val)
+	// We could do others here, but coercing from anything else is dangerous.
+	// Even uint could be 64 bits, though in practice nothing we work with here
+	// is.
+	default:
+		return nil, fmt.Errorf("unable to marshal %T to uint32", v)
+	}
+
```

Comment:
Can you explain why we need to add the new uint type support?

Reply:
Oh, sure. We were getting types unmarshalled (specifically uint) that didn't correspond to straight uint32 and it was panicking because it's not even a checked type assertion, and Go considers uint to be different from uint32 (even on 32-bit platforms, I believe). This at least covers the reasonable subset that can be expected/safely coerced, and returns an error if it's not compatible instead of panicking.

It's been a bit since I made the change, so I can't quite quote chapter and verse where the issue originated, though I did manage to track it down before. My vague recollection is that the unmarshalling code unmarshals to a uint, but I don't remember why I didn't fix it there (perhaps I tried it and it broke a lot of other things, don't recall).

Reply:
It may actually have specifically been this bit with TimeTicks: https://github.com/soniah/gosnmp/blob/2154f1e7f67afc47eb36b0bff4fc5b4d4e952cbf/helper.go#L581

Reply:
And, looking back at it, I'm not 100% sold on it being the correct way to do it, since `parseUint()` actually parses a `uint64` and does a quick check to see if it's too big to be cast as a `uint` (which may vary depending on whether it's a 32-bit or 64-bit arch) and barfs otherwise, while TimeTicks should actually explicitly be a `uint32`. Perhaps I should make the fix there instead.